### DISCUSSION TITLE
reopen_when seems to only be called every 2 minutes

### DISCUSSION THREAD

**Discussion by TheFrankman** [savonet/liquidsoap#2838] [User Support]
Hello There,

I can't figure out how to use the mailing list thing so i'm trying here.

I have found a similar issue in the mailing list thread here https://sourceforge.net/p/savonet/mailman/message/37218908/

I was attempting to use the following code : 

```
output.file(%mp3, '/my/path/%Y-%m-%d/%Y-%m-%d-%H_%M_%S.mp3', stream, reopen_when={0s or 10s or 20s or 30s or 40s or 50s}, fallible=true, on_stop=shutdown)
```

But it seems to only re-open a file very 2 minutes.

I then tried something a bit more complicated using a thread : 

```
reopen = ref(false)

def reopen_when() =
  log.debug(label="testing", "REOPEN_WHEN ASKED")
  if !reopen then
    log.debug(label="testing", "REOPEN_WHEN IS TRUE")
    reopen := false
    true
  else
    log.debug(label="testing", "REOPEN_WHEN IS FALSE")
    false
  end
end

def set_reopen(value) =
  log.debug(label="testing", "SET REOPEN CALLED : #{value}")
  reopen := value
end

s = output.file(%mp3(bitrate=128, id3v2=true),
            fallible=true,
            reopen_on_metadata=false,
            '/my/path/%Y-%m-%d/%Y-%m-%d-%H_%M_%S.mp3',
            stream, reopen_when=reopen_when)

thread.run(every=10., { set_reopen(true) })

```

I get the same result with this method. So it seems there is some kind of weird functionality on the reload_when function that I'm not understanding.

From the logs below, you can see that my thread is successfully setting the value of reopen every 10s but only every 2 minutes does the log from the reopen_when function appear

```
023/01/09 16:05:10 [testing:5] SET REOPEN CALLED : true
2023/01/09 16:05:20 [testing:5] SET REOPEN CALLED : true
[mp3float @ 0xffff94069c70] Could not update timestamps for skipped samples.
2023/01/09 16:05:30 [testing:5] SET REOPEN CALLED : true
[mp3float @ 0xffff9420a3c0] Could not update timestamps for skipped samples.
2023/01/09 16:05:40 [testing:5] SET REOPEN CALLED : true
2023/01/09 16:05:50 [testing:5] REOPEN_WHEN ASKED
2023/01/09 16:05:50 [testing:5] REOPEN_WHEN IS TRUE
2023/01/09 16:05:50 [testing:5] SET REOPEN CALLED : true
2023/01/09 16:06:00 [testing:5] SET REOPEN CALLED : true
[mp3float @ 0xffff94147d20] Could not update timestamps for skipped samples.
2023/01/09 16:06:10 [testing:5] SET REOPEN CALLED : true
[mp3float @ 0xffff94265e90] Could not update timestamps for skipped samples.
2023/01/09 16:06:20 [testing:5] SET REOPEN CALLED : true
2023/01/09 16:06:30 [testing:5] SET REOPEN CALLED : true
[mp3float @ 0xffff940e6660] Could not update timestamps for skipped samples.
2023/01/09 16:06:40 [testing:5] SET REOPEN CALLED : true
2023/01/09 16:06:50 [testing:5] SET REOPEN CALLED : true
[mp3float @ 0xffff94113f20] Could not update timestamps for skipped samples.
2023/01/09 16:07:00 [testing:5] SET REOPEN CALLED : true
2023/01/09 16:07:10 [testing:5] SET REOPEN CALLED : true
[mp3float @ 0xffff9426fb60] Could not update timestamps for skipped samples.
2023/01/09 16:07:20 [testing:5] SET REOPEN CALLED : true
2023/01/09 16:07:30 [testing:5] SET REOPEN CALLED : true
[mp3float @ 0xffff9426fb60] Could not update timestamps for skipped samples.
2023/01/09 16:07:40 [testing:5] SET REOPEN CALLED : true
2023/01/09 16:07:50 [testing:5] REOPEN_WHEN ASKED
2023/01/09 16:07:50 [testing:5] REOPEN_WHEN IS TRUE
2023/01/09 16:07:50 [testing:5] SET REOPEN CALLED : true
[mp3float @ 0xffff940931c0] Could not update timestamps for skipped samples.
2023/01/09 16:08:00 [testing:5] SET REOPEN CALLED : true
2023/01/09 16:08:10 [testing:5] SET REOPEN CALLED : true
[mp3float @ 0xffff941716b0] Could not update timestamps for skipped samples.
2023/01/09 16:08:20 [testing:5] SET REOPEN CALLED : true
```

**Comment by TheFrankman** [comment]
Even if you do something like this, it's still only re-opening every 2 minutes : 

```
output.file(
    %mp3,
    '/my/path/%Y-%m-%d/%Y-%m-%d-%H_%M_%S.mp3',
    stream,
    reopen_when={ int_of_float(gettimeofday()/60.) mod 1 == 0 },
    fallible=true,
    on_stop=shutdown
)
```

**Comment by vitoyucepi** [comment]
Hi @TheFrankman,
I think this works in another way.
`output.file` has parameters
1. `reopen_delay`
   > `reopen_delay` (of type `float`, which defaults to `120.`): Prevent re-opening within that delay, in seconds.

   In your case, you have to set it to `10.` seconds.
1. `reopen_when`
   > `reopen_when` (of type `() -> bool`, which defaults to `{false}`): When should the output be re-opened.

Every `reopen_delay` seconds `output.file` will call `reopen_when`.
If it returns `true`, then it will reopen the file.

Why do you want to cut every 10 seconds?

**Comment by cdgraff** [comment]
Hi @TheFrankman I was how sent that to the List, I change the file to hls as I need to be able to concat later portions of the files, created, and FILE doesn't work for that... can explain more about your use case. Possibly I can give other advice...

**Comment by TheFrankman** [comment]
I am just trying to create a backup in 10 second chunks with an aim to concatenate, I have since switched it to HLS and then just disregard the m3u8 file BUT... It feels like the `reopen_when` method is not properly documented as if you read the section in the book, or look at the online documentation it seems as though you should be able to use `reopen_when` I have also seen @toots say this should be possible :/

**Comment by vitoyucepi** [comment]
> It feels like the reopen_when method is not properly documented as if you read the section in the book, or look at the online documentation it seems as though you should be able to use reopen_when

The documentation could be out of sync with the current implementation.
Have a look at https://www.liquidsoap.info/doc-2.1.3/reference.html#output.file.

Previously you said

> every 2 minutes

> it's still only re-opening every 2 minutes

And it seems that `reopen_delay` is exactly what you want to tune!

```ruby
output.file(
    id='my_output',
    fallible=true,
    reopen_on_metadata=false,
    reopen_delay=10.0,
    reopen_when={true},
    %mp3(bitrate=128, id3v2=true),
    '/tmp/%Y-%m-%d/%Y-%m-%d-%H_%M_%S.mp3',
    blank()
)
```
I modified your code sample
1. `s =` is useless because `output.file` returns nothing.
1. `id='my_output'` custom id because the default one looks awful.
1. `reopen_delay=10.0` reopen every 10 seconds.
1. `reopen_when={true}` always returns `true`.
1. `blank()` sample stream.
1. `'/tmp/%Y-%m-%d/%Y-%m-%d-%H_%M_%S.mp3'` output to `/tmp`.

The output files look as expected
```
2023-01-10-13_06_31.mp3
2023-01-10-13_06_41.mp3
2023-01-10-13_06_51.mp3
2023-01-10-13_07_01.mp3
2023-01-10-13_07_11.mp3
```
And in the logs
```
2023/01/10 13:06:41 [my_output:3] Re-opening output pipe.
2023/01/10 13:06:51 [my_output:3] Re-opening output pipe.
2023/01/10 13:07:01 [my_output:3] Re-opening output pipe.
2023/01/10 13:07:11 [my_output:3] Re-opening output pipe.
```

P.S. https://www.liquidsoap.info/doc-dev/migrating.html#strftime

**Comment by cdgraff** [comment]
Good catch @vitoyucepi the reopen_when={true} I don't know that was possible I resolved these way:
```
    output.file(%mp3(bitrate=64, samplerate=44100, stereo=true),
       fallible=true,
       on_close=on_file_close,
       {"#{storage_dir}/#{station_slug}/%Y/%m/%d/%Hh/#{int_of_float(time())}.mp3"},
        reopen_delay=1.0,
        reopen_when={1s or 31s},
        cleanlive)
```

**Comment by cdgraff** [comment]
But IMPORTANT, that audios can't be concatenated back to be used in a single mp3 file, as split using `output.file` doesn't close clean, and if you try to concat them will have noise into the transition and FFMPEG will show errors in the reading process.

**Comment by toots** [comment]
I believe that if you ask ffmpeg to repack it, it should work. Something like:
```
ffmpeg -f concat -safe 0 -i <file list> -c copy out.mp3
```
What won't work is a straight byte/file copy:
```
cat <file list> > out.mp3
```

**Comment by cdgraff** [comment]
I tried with

`ffmpeg -i "concat:1669758001.mp3|1669758031.mp3|1669758061.mp3" -acodec copy output.mp3`

`ffmpeg -i "concat:1669758001.mp3|1669758031.mp3|1669758061.mp3" -acodec mp3 output.mp3`

Using COPY is worst than transcoding again to MP3, but if you listen depending on the sound you in both cases can see a missing ms of audio in the merge or noise into the transition, I show @toots into Slack many test and analyses with waveform and others... and later you advice me to move to HLS to warranty the merge quality, and I agree, HLS works perfect, and after many settings never reach the same results using output.file

**Comment by TheFrankman** [comment]
Thanks guys, I think this is more than enough to go on for me. In my instance I think that outputting using HLS is actually the better method as it's easier to concat a sub-section of the recordings. I was just frustrated that the reopen_when didn't seem to function as the documentation / book would suggest.

Appreciate all your help.
